src/common/mem.c

```diff
@@ -2195,29 +2491,592 @@ umem_cache_checkpoint(struct umem_store *store, umem_cache_wait_cb_t wait_cb, vo
+{
+	struct umem_page_info	*pinfo;
+	uint32_t		 pg_id;
+	int			 i, processed = 0, rc = 0;
+
+	for (i = 0; i < page_nr; i++) {
+		pg_id = pages[i];
+		pinfo = cache->ca_pages[pg_id].pg_info;
+
+		/* The page is already mapped */
+		if (pinfo != NULL) {
+			D_ASSERT(pinfo->pi_pg_id == pg_id);
+			D_ASSERT(pinfo->pi_mapped == 1);
+			cache_pin_page(cache, pinfo);
+			processed++;
+			continue;
+		}
+
+		rc = cache_get_free_page(cache, &pinfo, processed, for_sys);
+		if (rc)
+			goto error;
+
+		cache_map_page(cache, pinfo, pg_id);
+		cache_pin_page(cache, pinfo);
+		processed++;
+	}
+
+	for (i = 0; i < page_nr; i++) {
+		pg_id = pages[i];
+		pinfo = cache->ca_pages[pg_id].pg_info;
+
+		D_ASSERT(pinfo != NULL);
+		if (pinfo->pi_loaded == 0) {
+			rc = cache_load_page(cache, pinfo);
+			if (rc) {
+				goto error;
+			} else if (pinfo->pi_loaded == 0) {
+				rc = -DER_IO;
+				goto error;
+			}
+		}
+		pinfo->pi_sys = for_sys;
+	}
+
+	return 0;
+error:
+	for (i = 0; i < processed; i++) {
+		pg_id = pages[i];
+		pinfo = cache->ca_pages[pg_id].pg_info;
+
+		D_ASSERT(pinfo != NULL);
+		cache_unpin_page(cache, pinfo);
+
+	}
+	return rc;
+}
+
+#define DF_RANGE		\
+	DF_U64", "DF_U64
+#define DP_RANGE(range)		\
+	(range)->cr_off, (range)->cr_size
+
+static int
+cache_rgs2pgs(struct umem_cache *cache, struct umem_cache_range *ranges, int range_nr,
+	      uint32_t *in_pages, int *page_nr, uint32_t **out_pages)
+{
+	struct umem_cache_range	range;
+	uint32_t		page_id, *pages = in_pages, *old_pages = NULL, len = 0;
+	int			rc = 0, i, page_idx = 0, tot_pages = *page_nr;
+
+	for (i = 0; i < range_nr; i++) {
+		range = ranges[i];
+		/* Assume the ranges are sorted & no overlapping */
+		if (i > 0) {
+			if (range.cr_off < ranges[i - 1].cr_off + ranges[i - 1].cr_size) {
+				D_ERROR("Invalid ranges ["DF_RANGE"], ["DF_RANGE"]\n",
+					DP_RANGE(&ranges[i - 1]), DP_RANGE(&range));
+				rc = -DER_INVAL;
+				goto error;
+			}
+		}
+
+		D_ASSERT(range.cr_size > 0);
+		while (range.cr_size > 0) {
+			page_id = range.cr_off >> cache->ca_page_shift;
+
+			if (len != 0 && page_id != pages[page_idx]) {
+				page_idx++;
+				if (page_idx == tot_pages) {
+					D_REALLOC_ARRAY(pages, old_pages, tot_pages, tot_pages * 2);
+					if (pages == NULL) {
+						D_ERROR("Alloc array(%d) failed.\n", tot_pages * 2);
+						rc = -DER_NOMEM;
+						goto error;
+					}
+					old_pages = pages;
+					tot_pages = tot_pages * 2;
+				}
+			}
+
+			pages[page_idx] = page_id;
+			len = cache->ca_page_sz - (range.cr_off & cache->ca_page_mask);
+			range.cr_off += len;
+			if (range.cr_size >= len)
+				range.cr_size -= len;
+			else
+				range.cr_size = 0;
+		}
+	}
+
+	D_ASSERT(page_idx < tot_pages);
+	*out_pages = pages;
+	*page_nr = page_idx + 1;
+
+	return 0;
+error:
+	if (old_pages)
+		D_FREE(old_pages);
+	return rc;
+}
+
+#define UMEM_PAGES_ON_STACK	16
+
+int
+umem_cache_map(struct umem_store *store, struct umem_cache_range *ranges, int range_nr)
```

Comment:
umem_cache_map() can be called for evictable pages as-well. This can be the case till all of the md blob zones are used. This is not a must, but good to have requirement.

Reply:
The question is that is allocator able to tell if a bucket is evictable or not before it's loaded? I think I've asked this before, and your answer is that allocator needs to load the bucket first, otherwise it won't tell if the bucket is evictable or not.

I "think" allocator should be able to tell if a bucket is evict-able or not (if it put some necessary global information in non-evictable zone). If umem cache is able to tell if a bucket is evict-able by ID (without loading it), it could simplify current implementation a bit.